ddl/index.go

```diff
@@ -61,6 +67,43 @@ func buildIndexInfo(tblInfo *model.TableInfo, unique bool, indexName model.CIStr
 			return nil, errors.Trace(errTooLongKey)
 		}
 
+		// take care of the sum of length of all index columns
+		if ic.Length != types.UnspecifiedLength {
+			sumMaxLength += ic.Length
+		} else {
+			// specified data types
+			if col.Flen != types.UnspecifiedLength {
+				// special case for the bit type
+				if col.FieldType.Tp == mysql.TypeBit {
+					sumMaxLength += int(math.Ceil(float64(col.Flen+7) / float64(8)))
+				} else {
+					sumMaxLength += col.Flen
+				}
+			} else {
+				if len, ok := mysql.DefaultLengthOfMysqlTypes[col.FieldType.Tp]; ok {
+					sumMaxLength += len
+				} else {
+					return nil, errUnknownTypeLength.GenByArgs(col.FieldType.Tp)
+				}
+
+				// special case for time fraction
+				if (col.FieldType.Tp == mysql.TypeDatetime ||
+					col.FieldType.Tp == mysql.TypeDuration ||
+					col.FieldType.Tp == mysql.TypeTimestamp) && col.FieldType.Decimal != -1 {
+
+					if len, ok := mysql.DefaultLengthOfTimeFraction[col.FieldType.Decimal]; ok {
+						sumMaxLength += len
+					} else {
+						return nil, errUnknownFractionLength.GenByArgs(col.FieldType.Tp, col.FieldType.Decimal)
```

Comment:
Can the code of line71-line101 be extracted into a function?
This code is nested too long and it makes the code line of `BuildIndexInfo` too long.